tests/rptest/tests/pandaproxy_test.py

```diff
@@ -2231,3 +2243,234 @@ def do_coordinator_change(utils: GroupCoordinatorTransferUtils):
         self.logger.debug(
             "Test offset commit: after second coordinator change")
         do_consumer_offset_commit(new_offset=1)
+
+
+class VerifiablePriorityQueue:
+    """
+    A priority queue used with VerifiablePandaproxyProducers/Consumers to verify some event properties:
+    (1) Every event produced is also consumed
+    (2) Every event is consumed atleast once
+    (3) Events are consumed in the same order they are produced in all consumers
+
+    This class uses python's PriorityQueue which internally use locks to temporarily block competing threads.
+    See python docs for more info: https://docs.python.org/3/library/queue.html
+    """
+    def __init__(self,
+                 logger: TestLoggerMaker,
+                 maxsize: int = 0,
+                 q_name: str = gen_ascii(16)):
+        self.last_known_pos = -1
+        self.q = PriorityQueue(maxsize)
+        self.q_name = q_name
+        self.logger = logger
+
+    def has_events(self):
+        return self.q.qsize() > 0
+
+    def add(self, timestamp: datetime.datetime, event: Any):
+        timestamp_event = (timestamp, event)
+        self.logger.debug(
+            f"Queue {self.q_name}, Adding timestamp_event {timestamp_event}")
+        self.q.put(timestamp_event)
+        self.logger.debug(
+            f"Queue {self.q_name}, relative size {self.q.qsize()} after add")
+
+    def remove(self):
+        self.logger.debug(f"Queue {self.q_name}, Removing timestamp_event...")
+        timestamp_event = self.q.get()
+        self.logger.debug(
+            f"Queue {self.q_name}, Removed timestamp_event {timestamp_event}, relative size {self.q.qsize()}"
+        )
+        return timestamp_event
+
+    def get_event_pos(self, event: Any, consumers: list):
+        assert len(consumers) > 0
+
+        def do_get_event_pos():
+            try:
+                i = consumers[0].events.index(event)
```

Comment:
So the test successfully handles changes in topic-partition leadership by re-issuing the fetch request. So that goal is reached; however, the verification process keeps failing here because apparently the consumer never fetched an event 

```
[INFO:2023-04-28 09:28:21,198]: starting test run with session id 2023-04-28--020...
[INFO:2023-04-28 09:28:21,198]: running 1 tests...
[INFO:2023-04-28 09:28:21,198]: Triggering test 1 of 1...
[INFO:2023-04-28 09:28:21,202]: RunnerClient: Loading test {'directory': '/root/tests/rptest/tests', 'file_name': 'pandaproxy_test.py', 'cls_name': 'LeadershipTransferTest', 'method_name': 'test_leadership_transfer_with_load', 'injected_args': None}
[INFO:2023-04-28 09:28:21,205]: RunnerClient: rptest.tests.pandaproxy_test.LeadershipTransferTest.test_leadership_transfer_with_load: Setting up...
[INFO:2023-04-28 09:28:28,053]: RunnerClient: rptest.tests.pandaproxy_test.LeadershipTransferTest.test_leadership_transfer_with_load: Running...
Exception in thread Thread-9 (remove_from_q):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/root/tests/rptest/tests/pandaproxy_test.py", line 2563, in remove_from_q
    self.prio_q.verify_event_in_consumers(timestamp_event[1],
  File "/root/tests/rptest/tests/pandaproxy_test.py", line 2335, in verify_event_in_consumers
    event_pos = self.get_event_pos(event, consumers)
  File "/root/tests/rptest/tests/pandaproxy_test.py", line 2298, in get_event_pos
    wait_until_result(
  File "/root/tests/rptest/util.py", line 90, in wait_until_result
    wait_until(wrapped_condition, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ducktape/utils/util.py", line 57, in wait_until
    raise TimeoutError(err_msg() if callable(err_msg) else err_msg) from last_exception
ducktape.errors.TimeoutError: Failed to get event position. No consumer fetched the event.
[WARNING - 2023-04-28 09:29:32,176 - redpanda - raise_on_bad_logs - lineno:2016]: Ignoring allow-listed log line 'ERROR 2023-04-28 09:28:23,847 [shard 0] syschecks - Path: `/var/lib/redpanda/data' is not on XFS or ext4. This is a non-supported configuration. You may experience poor performance or instability.'
[WARNING - 2023-04-28 09:29:32,223 - redpanda - raise_on_bad_logs - lineno:2016]: Ignoring allow-listed log line 'ERROR 2023-04-28 09:28:23,847 [shard 0] syschecks - Path: `/var/lib/redpanda/data' is not on XFS or ext4. This is a non-supported configuration. You may experience poor performance or instability.'
[WARNING - 2023-04-28 09:29:32,274 - redpanda - raise_on_bad_logs - lineno:2016]: Ignoring allow-listed log line 'ERROR 2023-04-28 09:28:23,847 [shard 0] syschecks - Path: `/var/lib/redpanda/data' is not on XFS or ext4. This is a non-supported configuration. You may experience poor performance or instability.'
[INFO:2023-04-28 09:29:32,318]: RunnerClient: rptest.tests.pandaproxy_test.LeadershipTransferTest.test_leadership_transfer_with_load: PASS
[INFO:2023-04-28 09:29:32,319]: RunnerClient: rptest.tests.pandaproxy_test.LeadershipTransferTest.test_leadership_transfer_with_load: Tearing down...
[INFO:2023-04-28 09:29:35,132]: RunnerClient: rptest.tests.pandaproxy_test.LeadershipTransferTest.test_leadership_transfer_with_load: Summary: 
[INFO:2023-04-28 09:29:35,132]: RunnerClient: rptest.tests.pandaproxy_test.LeadershipTransferTest.test_leadership_transfer_with_load: Data: None
test_id:    rptest.tests.pandaproxy_test.LeadershipTransferTest.test_leadership_transfer_with_load
status:     PASS
run time:   1 minute 13.926 seconds
--------------------------------------------------------------------------------
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.8.8
session_id:       2023-04-28--020
run time:         1 minute 13.939 seconds
tests run:        1
passed:           1
failed:           0
ignored:          0
opassed:          0
ofailed:          0
================================================================================
```

I'm struggling to figure out why this is happening but perhaps I need to sleep on it.
From the logs, we see that this timeout happens on the first call to remove an item from the priority queue. We also observe that the producer successfully adds to the queue many times before the first call to remove an event. So, off the top of my head, I'm wondering if pythons priority queue (or perhaps locking mechs) give preference to the first "thing" that access it. In the test code, the producer is run first so it likely writes to the queue first.

Reply:
Look at the logs some more,
I see that the first produce event was
```
[DEBUG - 2023-04-28 09:28:28,307 - pandaproxy_test - add - lineno:2273]: Queue oxlevghwastbnqnj, Adding timestamp_event (datetime.datetime(2023, 4, 28, 9, 28, 28, 307416), {'topic': 'topic-raghjkszch', 'key': None, 'value': 'pvretqsvhgzzkwnnjkpipkourihdhstjnymerowyjoikkeczbogaabuzdchluauyobyimftpdaebnqsltxhfkzbagcnnwsrbcdbwwjhowqfdisvoqafiuhgboxzauczylfmgxwwhoqncyohgxudzmoyidrdcafhtacgqsvyazkpvhjcbxrsqtqqkdabijzanaklxcoeqfspuaimaftmkujwpfvxewqadlktodqioxrilavojawtyprwyrwmatsudijoczmdopjowitknuthlqzlnvgbpqwxotayquaojwzmsbmtaphnhdbaegrgnsqpoadczlgxmvyvmfrsigllittuiwlepxixjuywcqdtksyhopmwphuetpvhxemwbiytzeajwzepxlmpxobnevczdsjxibjixurijdtmobsevnibzhojbxfgxnwkbyxbmfbrdfpxeghfielkeczgjwdjxltcoakjrnlghwwnykbnpzehapmwofgonkgnfycbvwfbq', 'partition': 0, 'offset': 0})
```

So that should be the first event the log; however, the fetch result from the consumer does not have that value even though PP consumer always consumer from the beginning of the topic. This may be a bug.